Realm/RLMMongoCollection.mm

```diff
@@ -0,0 +1,307 @@
+    });
+}
+
+- (void)updateOneDocumentWhere:(NSDictionary<NSString *, id<RLMBSON>> *)filterDocument
+                updateDocument:(NSDictionary<NSString *, id<RLMBSON>> *)updateDocument
+                    completion:(RLMMongoUpdateBlock)completion {
+    [self updateOneDocumentWhere:filterDocument
+                  updateDocument:updateDocument
+                          upsert:NO
+                      completion:completion];
+}
+
+- (void)updateManyDocumentsWhere:(NSDictionary<NSString *, id<RLMBSON>> *)filterDocument
+                  updateDocument:(NSDictionary<NSString *, id<RLMBSON>> *)updateDocument
+                          upsert:(BOOL)upsert
+                      completion:(RLMMongoUpdateBlock)completion {
+    [self collection:self.name].update_many(static_cast<realm::bson::BsonDocument>(RLMConvertRLMBSONToBson(filterDocument)),                 static_cast<realm::bson::BsonDocument>(RLMConvertRLMBSONToBson(updateDocument)),
+                                            upsert,
+                                            [completion](realm::app::RemoteMongoCollection::RemoteUpdateResult result,
+                                                         realm::util::Optional<realm::app::AppError> error) {
+        if (error) {
+            return completion(nil, RLMAppErrorToNSError(*error));
+        }
+        completion([[RLMUpdateResult alloc] initWithRemoteUpdateResult:result], nil);
+    });
+}
+
+- (void)updateManyDocumentsWhere:(NSDictionary<NSString *, id<RLMBSON>> *)filterDocument
+                  updateDocument:(NSDictionary<NSString *, id<RLMBSON>> *)updateDocument
+                      completion:(RLMMongoUpdateBlock)completion {
+    [self updateManyDocumentsWhere:filterDocument
+                    updateDocument:updateDocument
+                            upsert:NO
+                        completion:completion];
+}
+
+- (void)findOneAndUpdateWhere:(NSDictionary<NSString *, id<RLMBSON>> *)filterDocument
+               updateDocument:(NSDictionary<NSString *, id<RLMBSON>> *)updateDocument
+                      options:(RLMFindOneAndModifyOptions *)options
+                   completion:(RLMMongoFindOneBlock)completion {
+    [self collection:self.name].find_one_and_update(static_cast<realm::bson::BsonDocument>(RLMConvertRLMBSONToBson(filterDocument)), static_cast<realm::bson::BsonDocument>(RLMConvertRLMBSONToBson(updateDocument)),
+                                                    [options _findOneAndModifyOptions],
+                                                    [completion](realm::util::Optional<realm::bson::BsonDocument> document,
+                                                                 realm::util::Optional<realm::app::AppError> error) {
+        if (error) {
+            return completion(nil, RLMAppErrorToNSError(*error));
+        }
+
+        if (document) {
+            return completion((NSDictionary *)RLMConvertBsonToRLMBSON(*document), nil);
```

Comment:
I notice that we do the equivalent of `foo ? RLMConvertBsonToRLMBSON(*foo) : nil` in an awful lot of places, which suggests that a RLMConvertBsonToRLMBSON overload which takes an optional directly would eliminate a lot of repetition.

Reply:
This suggestion would be good for a future improvements PR, @jsflax

Reply:
It's a fairly simple addition if you want to add it:

```objc
id<RLMBSON> RLMConvertBsonToRLMBSON(realm::util::Optional<Bson> b) {
    return b ? RLMConvertBsonToRLMBSON(*b) : nil;
}
```

or

```objc
id<RLMBSON> RLMConvertBsonToRLMBSON(realm::util::Optional<Bson> b) {
    return RLMConvertBsonToRLMBSON(b.value_or(Bson()));
}
```